components/homme/src/share/cxx/scream_kokkos_utils.hpp

```diff
@@ -0,0 +1,189 @@
+#ifndef SCREAM_KOKKOS_UTILS_HPP
+#define SCREAM_KOKKOS_UTILS_HPP
+
+#include "Types.hpp"
+#include "ExecSpaceDefs.hpp"
+
+#include "../../algorithms/src/Kokkos_Random.hpp"
+
+#include <chrono>
+#include <cassert>
+#include <type_traits>
+
+// This file should not be merged with share/scream_kokkos_meta.hpp.
+// That file contains functionalities that *ideally* should be in kokkos
+// itself, and we can foresee being in kokkos in the future. Hence, that
+// file may be gone at some point (in fact, it may be gone by the time
+// you read this comment). This file, instead, contains functionalities
+// that are probably not generic enough to appear in kokkos any time soon
+// (or ever), and are more app-specific.
+
+namespace Homme {
+
+/*
+ * TeamUtils contains utilities for getting concurrency info for
+ * thread teams. Don't use _TeamUtilsCommonBase directly, use
+ * TeamUtils.
+ */
+
+template <typename ExeSpace=ExecSpace>
+class _TeamUtilsCommonBase
+{
+ protected:
+  int _team_size, _num_teams, _max_threads, _league_size;
+
+ public:
+  template <typename TeamPolicy>
+  _TeamUtilsCommonBase(const TeamPolicy& policy)
+  {
+    _max_threads = ExeSpace::concurrency() / ( OnGpu<ExeSpace>::value ? 2 : 1);
+    const int team_size = policy.team_size();
+    _num_teams = _max_threads / team_size;
+    _team_size = _max_threads / _num_teams;
+    _league_size = policy.league_size();
+
+    // We will never run more teams than the policy needs
+    _num_teams = _num_teams > _league_size ? _league_size : _num_teams;
+
+    assert(_num_teams > 0);
+  }
+
+  // How many thread teams can run concurrently
+  int get_num_concurrent_teams() const { return _num_teams; }
+
+  // How many threads can run concurrently
+  int get_max_concurrent_threads() const { return _max_threads; }
+
+  // How many ws slots are there
+  int get_num_ws_slots() const { return _num_teams; }
+
+  /*
+   * Of the C concurrently running teams, which "slot" is open
+   * for the given team.
+   */
+  template <typename MemberType>
+  KOKKOS_INLINE_FUNCTION
+  int get_workspace_idx(const MemberType& /*team_member*/) const
+  { return 0; }
+
+  template <typename MemberType>
+  KOKKOS_INLINE_FUNCTION
+  void release_workspace_idx(const MemberType& /*team_member*/, int /*ws_idx*/) const
+  { }
+};
+
+template <typename ExeSpace = Kokkos::DefaultExecutionSpace>
+class TeamUtils : public _TeamUtilsCommonBase<ExeSpace>
```

Comment:
Design choice, feel free to push back: do we need the inheritance pattern here? There's no extension of the base class, nor virtual functions (because Cuda). It looks to me as this class should simply be a template class specialized later on, without a base class (there's close to 0 code reuse anyways).

Second thought: can we leave only the template class in this hpp file, and relegate the specializations to a cpp file, with ETI? That way we would keep the interface (that one needs to be aware of) in a small tidy file, rather than a long one. Or would that require relocatable-device-code? (I admit that I'm a noob when it comes to understand rdc mechanism).

Reply:
I think the "second thought" would require rdc, so you can disregard that comment.

The thought about simple specialization vs inheritance is the only one to focus here.

Reply:
I'm tempted to keep it since we are getting some code re-use in the shared constructor and in some of the queries.